<a href="https://colab.research.google.com/github/Samujjalborah/python-learning-journey/blob/main/Day16/Day16_Student_Wellbeing_Statistical_Analysis_Probability.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Day 16 – Student Wellbeing Statistical Analysis & Probability

## Objective

This notebook performs statistical analysis and probability calculations on a Student Wellbeing Survey dataset.

The notebook includes:

- Dataset inspection
- Measures of central tendency
- Measures of dispersion
- Outlier detection using IQR
- Probability calculations
- Conditional probability
- Bayes' theorem
- Normal distribution analysis
- Statistical observations

Developed by: Samujjal Borah

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use("default")
df = pd.read_csv("Student_survey.csv")

df.head()

,Student_ID,Age,Faculty,Year_of_Study,City,Accommodation,Scholarship,Part_Time_Job,Internet_Quality,Preferred_Study_Space,Weekly_Study_Hours,Average_Sleep_Hours,Daily_Screen_Time_Hours,Exercise_Days_Per_Week,Commute_Time_Minutes,Stress_Score,Academic_Readiness_Score,Overall_Satisfaction,Monthly_Discretionary_Spending,Social_Activity_Hours_Per_Week
0,STU0001,21,Data Science,3,Bengaluru,Hostel,No,Yes,Good,Café,15.2,5.7,3.2,3,11.8,5.4,63.1,3.2,5131.0,7.8
1,STU0002,21,Data Science,4,Pune,Hostel,Yes,No,Poor,Library,17.6,7.7,3.5,4,20.3,2.1,74.2,3.8,3508.0,9.2
2,STU0003,22,Life Sciences,2,Chandigarh,Home,Yes,No,Poor,Home,18.9,6.4,2.8,1,15.0,5.2,77.3,4.0,4584.0,7.7
3,STU0004,18,Business,4,Bengaluru,Hostel,Yes,Yes,Average,Café,19.5,6.7,3.0,5,12.1,4.7,67.0,4.0,5456.0,5.3
4,STU0005,20,Business,3,Jammu,Shared Apartment,Yes,No,Good,Home,18.4,7.3,3.0,1,30.2,5.4,65.6,3.2,11732.0,15.9


Dataset Overview

In [2]:
print("="*60)
print("DATASET SHAPE")
print("="*60)
print(df.shape)

print("\n")

print("="*60)
print("FIRST FIVE RECORDS")
print("="*60)
display(df.head())

print("\n")

print("="*60)
print("DATA TYPES")
print("="*60)
display(df.dtypes)

print("\n")

print("="*60)
print("MISSING VALUES")
print("="*60)

missing = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Percentage": round(df.isnull().mean()*100,2)
})

display(missing[missing["Missing Values"]>0])

DATASET SHAPE
(600, 20)


FIRST FIVE RECORDS


,Student_ID,Age,Faculty,Year_of_Study,City,Accommodation,Scholarship,Part_Time_Job,Internet_Quality,Preferred_Study_Space,Weekly_Study_Hours,Average_Sleep_Hours,Daily_Screen_Time_Hours,Exercise_Days_Per_Week,Commute_Time_Minutes,Stress_Score,Academic_Readiness_Score,Overall_Satisfaction,Monthly_Discretionary_Spending,Social_Activity_Hours_Per_Week
0,STU0001,21,Data Science,3,Bengaluru,Hostel,No,Yes,Good,Café,15.2,5.7,3.2,3,11.8,5.4,63.1,3.2,5131.0,7.8
1,STU0002,21,Data Science,4,Pune,Hostel,Yes,No,Poor,Library,17.6,7.7,3.5,4,20.3,2.1,74.2,3.8,3508.0,9.2
2,STU0003,22,Life Sciences,2,Chandigarh,Home,Yes,No,Poor,Home,18.9,6.4,2.8,1,15.0,5.2,77.3,4.0,4584.0,7.7
3,STU0004,18,Business,4,Bengaluru,Hostel,Yes,Yes,Average,Café,19.5,6.7,3.0,5,12.1,4.7,67.0,4.0,5456.0,5.3
4,STU0005,20,Business,3,Jammu,Shared Apartment,Yes,No,Good,Home,18.4,7.3,3.0,1,30.2,5.4,65.6,3.2,11732.0,15.9




DATA TYPES


,0
Student_ID,object
Age,int64
Faculty,object
Year_of_Study,int64
City,object
Accommodation,object
Scholarship,object
Part_Time_Job,object
Internet_Quality,object
Preferred_Study_Space,object




MISSING VALUES


,Missing Values,Percentage


## Dataset Overview

- The dataset contains **600 records and 20 variables**.
- No missing values were detected in any column.
- The dataset contains a mixture of numerical and categorical variables.
- Numerical variables are stored in appropriate numeric formats (`int64` and `float64`), while descriptive variables are stored as `object`.
- The dataset is ready for statistical analysis without additional preprocessing.


# Measures of Central Tendency

These statistics describe the central value of a dataset.

### Mean
Mean = Σx / n

### Median
Median = Middle value after sorting the observations.

### Mode
Mode = Most frequently occurring value.

In [3]:
variables = [
    "Weekly_Study_Hours",
    "Average_Sleep_Hours",
    "Daily_Screen_Time_Hours",
    "Stress_Score",
    "Academic_Readiness_Score"
]

central_tendency = pd.DataFrame({
    "Mean": df[variables].mean(),
    "Median": df[variables].median(),
    "Mode": df[variables].mode().iloc[0]
})

print("="*60)
print("MEASURES OF CENTRAL TENDENCY")
print("="*60)

display(central_tendency)

MEASURES OF CENTRAL TENDENCY


,Mean,Median,Mode
Weekly_Study_Hours,15.691000,15.40,13.1
Average_Sleep_Hours,6.997500,7.00,7.0
Daily_Screen_Time_Hours,4.503000,4.20,3.5
Stress_Score,4.464500,4.50,4.7
Academic_Readiness_Score,71.773167,71.65,71.1


## Interpretation of Central Tendency

- Weekly Study Hours average **15.69 hours** per week, with a median of **15.40 hours**, indicating a fairly balanced distribution.
- Students sleep approximately **7 hours** per day on average, with the mean and median being almost identical.
- Daily Screen Time averages **4.50 hours**, suggesting moderate daily digital device usage.
- The average Stress Score is **4.46**, indicating a moderate stress level among students.
- Academic Readiness has the highest average value (**71.77**), showing generally good preparedness across the surveyed students.

# Measures of Dispersion

### Range
Range = Maximum − Minimum

### Variance
Variance = Σ(x − Mean)² / n

### Standard Deviation
Standard Deviation = √Variance

### Interquartile Range (IQR)

IQR = Q3 − Q1

Lower Bound = Q1 − 1.5 × IQR

Upper Bound = Q3 + 1.5 × IQR

In [4]:
dispersion = pd.DataFrame({
    "Range": df[variables].max() - df[variables].min(),
    "Variance": df[variables].var(),
    "Standard Deviation": df[variables].std(),
    "Q1": df[variables].quantile(0.25),
    "Q3": df[variables].quantile(0.75)
})

dispersion["IQR"] = dispersion["Q3"] - dispersion["Q1"]

print("="*60)
print("MEASURES OF DISPERSION")
print("="*60)

display(dispersion)

MEASURES OF DISPERSION


,Range,Variance,Standard Deviation,Q1,Q3,IQR
Weekly_Study_Hours,30.0,19.217047,4.383725,13.000,18.425,5.425
Average_Sleep_Hours,5.0,0.703383,0.838679,6.475,7.600,1.125
Daily_Screen_Time_Hours,11.2,3.469006,1.862527,3.200,5.400,2.200
Stress_Score,8.4,2.782494,1.668081,3.300,5.700,2.400
Academic_Readiness_Score,56.1,93.981967,9.694430,65.200,78.025,12.825


## Interpretation of Dispersion

- Academic Readiness Score shows the greatest variability, with a standard deviation of **9.69**, indicating the widest spread of scores among students.
- Average Sleep Hours has the smallest variability (SD = **0.84**), meaning most students have similar sleeping habits.
- Weekly Study Hours exhibit moderate variation (IQR = **5.43 hours**).
- Daily Screen Time Hours also vary considerably across students (IQR = **2.20 hours**).
- Stress Score has an IQR of **2.40**, indicating moderate differences in stress levels among students.

Variable with Highest Variability

In [5]:
print("Variable with Highest Standard Deviation:")
print(dispersion["Standard Deviation"].idxmax())

print()

print("Highest Standard Deviation:")
print(round(dispersion["Standard Deviation"].max(),3))

Variable with Highest Standard Deviation:
Academic_Readiness_Score

Highest Standard Deviation:
9.694


# Outlier Detection using IQR

Any observation below

Q1 − 1.5 × IQR

or above

Q3 + 1.5 × IQR

is considered an outlier.

In [6]:
outlier_columns = [
    "Weekly_Study_Hours",
    "Daily_Screen_Time_Hours",
    "Commute_Time_Minutes",
    "Monthly_Discretionary_Spending"
]

outlier_summary = []

for col in outlier_columns:

    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower) | (df[col] > upper)]

    outlier_summary.append({
        "Variable": col,
        "Outliers": len(outliers)
    })

outlier_summary = pd.DataFrame(outlier_summary)

print("="*60)
print("OUTLIER SUMMARY (IQR METHOD)")
print("="*60)

display(outlier_summary)

OUTLIER SUMMARY (IQR METHOD)


,Variable,Outliers
0,Weekly_Study_Hours,8
1,Daily_Screen_Time_Hours,18
2,Commute_Time_Minutes,4
3,Monthly_Discretionary_Spending,20


## Interpretation of Outlier Analysis

The IQR method was used to detect outliers in four numerical variables.

- **Weekly Study Hours** contains **8 outliers**, indicating that a few students study substantially more or less than the majority.
- **Daily Screen Time Hours** contains **18 outliers**, suggesting that some students spend exceptionally high or low amounts of time on digital devices.
- **Commute Time Minutes** contains **4 outliers**, meaning only a small number of students have unusually long or short commutes.
- **Monthly Discretionary Spending** contains **20 outliers**, the highest among the selected variables, indicating that spending habits vary considerably for a small group of students.

Mean vs Median Before & After Removing Outliers

In [7]:
col = "Monthly_Discretionary_Spending"

Q1 = df[col].quantile(0.25)
Q3 = df[col].quantile(0.75)

IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

clean_df = df[(df[col] >= lower) & (df[col] <= upper)]

comparison = pd.DataFrame({
    "Before":[
        df[col].mean(),
        df[col].median()
    ],
    "After":[
        clean_df[col].mean(),
        clean_df[col].median()
    ]
}, index=["Mean","Median"])

print("="*60)
print("BEFORE VS AFTER REMOVING OUTLIERS")
print("="*60)

display(comparison)

BEFORE VS AFTER REMOVING OUTLIERS


,Before,After
Mean,6343.778333,5973.718966
Median,5685.500000,5590.500000


## Effect of Removing Outliers

Monthly Discretionary Spending was selected to compare statistics before and after removing outliers.

- Mean before removing outliers: **6343.78**
- Mean after removing outliers: **5973.72**
- Median before removing outliers: **5685.50**
- Median after removing outliers: **5590.50**

The mean decreased more noticeably than the median after removing outliers, indicating that a small number of extremely high spending values were increasing the average. The median changed only slightly, demonstrating that it is more robust to outliers.

# Probability

### Probability Formula

P(A) = Number of favourable outcomes / Total number of outcomes

### Conditional Probability

P(A|B) = P(A ∩ B) / P(B)

In [8]:
# Define Events

A = df["Part_Time_Job"] == "Yes"
B = df["Stress_Score"] >= 7
C = df["Scholarship"] == "Yes"
D = df["Exercise_Days_Per_Week"] >= 3

# Probabilities

P_A = A.mean()
P_B = B.mean()
P_C = C.mean()
P_D = D.mean()

P_A_or_B = (A | B).mean()
P_A_and_B = (A & B).mean()

P_A_given_B = (A & B).sum() / B.sum()
P_B_given_A = (A & B).sum() / A.sum()

probability = pd.DataFrame({
    "Probability":[
        P_A,
        P_B,
        P_C,
        P_D,
        P_A_or_B,
        P_A_and_B,
        P_A_given_B,
        P_B_given_A
    ]
},
index=[
    "P(A)",
    "P(B)",
    "P(C)",
    "P(D)",
    "P(A or B)",
    "P(A and B)",
    "P(A|B)",
    "P(B|A)"
])

print("="*60)
print("PROBABILITY ANALYSIS")
print("="*60)

display(probability)

PROBABILITY ANALYSIS


,Probability
P(A),0.253333
P(B),0.075000
P(C),0.306667
P(D),0.593333
P(A or B),0.276667
P(A and B),0.051667
P(A|B),0.688889
P(B|A),0.203947


## Probability Analysis

The probabilities of the defined events are:

- P(A): Probability that a student has a part-time job = **0.253**
- P(B): Probability that a student has a Stress Score of at least 7 = **0.075**
- P(C): Probability that a student has a scholarship = **0.307**
- P(D): Probability that a student exercises at least 3 days per week = **0.593**
- P(A ∪ B): Probability that a student has a part-time job or a high stress score = **0.277**
- P(A ∩ B): Probability that a student has both a part-time job and a high stress score = **0.052**
- P(A|B): Among students with high stress, approximately **68.9%** have a part-time job.
- P(B|A): Among students with part-time jobs, approximately **20.4%** have high stress.

Mutually Exclusive Events

In [9]:
year1 = (df["Year_of_Study"] == 1)
year4 = (df["Year_of_Study"] == 4)

mutual = (year1 & year4).sum()

print("Students in both Year 1 and Year 4:", mutual)

if mutual == 0:
    print("Conclusion: Year 1 and Year 4 are mutually exclusive events.")
else:
    print("Conclusion: They are NOT mutually exclusive.")

Students in both Year 1 and Year 4: 0
Conclusion: Year 1 and Year 4 are mutually exclusive events.


## Mutually Exclusive Events

Year of Study = 1 and Year of Study = 4 are **mutually exclusive events** because a student cannot belong to both years simultaneously. The intersection of these two events is zero.

Independence Check

In [10]:
print("P(A and B) =", round(P_A_and_B,4))
print("P(A) × P(B) =", round(P_A * P_B,4))

if abs(P_A_and_B - (P_A * P_B)) < 0.01:
    print("\nConclusion: Events A and B appear approximately independent.")
else:
    print("\nConclusion: Events A and B do not appear independent.")

P(A and B) = 0.0517
P(A) × P(B) = 0.019

Conclusion: Events A and B do not appear independent.


## Independence of Events A and B

The probability of both events occurring together is:

- P(A ∩ B) = **0.0517**

The product of their individual probabilities is:

- P(A) × P(B) = **0.0190**

Since these values are not equal, Events A (having a part-time job) and B (Stress Score ≥ 7) do **not** appear to be independent. This suggests that having a part-time job may be associated with students experiencing higher stress levels in this dataset.

#Bayes' Theorem

In [11]:
P_not_A = 1 - P_A

P_B_given_not_A = ((~A) & B).sum() / (~A).sum()

P_A_given_B_bayes = (P_B_given_A * P_A) / P_B

print("="*60)
print("BAYES' THEOREM")
print("="*60)

print("P(B|A):", round(P_B_given_A,4))
print("P(B|not A):", round(P_B_given_not_A,4))
print("P(A):", round(P_A,4))
print("P(B):", round(P_B,4))

print("\nP(A|B) using Bayes' Theorem:", round(P_A_given_B_bayes,4))
print("Direct Conditional Probability:", round(P_A_given_B,4))

BAYES' THEOREM
P(B|A): 0.2039
P(B|not A): 0.0312
P(A): 0.2533
P(B): 0.075

P(A|B) using Bayes' Theorem: 0.6889
Direct Conditional Probability: 0.6889


## Bayes' Theorem

Bayes' Theorem was used to calculate the probability that a student has a part-time job given that the student has a Stress Score of at least 7.

- P(B|A) = **0.2039**
- P(B|not A) = **0.0312**
- P(A) = **0.2533**
- P(B) = **0.0750**

Using Bayes' Theorem:

P(A|B) = **0.6889**

The directly calculated conditional probability is also:

P(A|B) = **0.6889**

Both values are identical, confirming that the Bayes' theorem calculation is correct.

#Normal Distribution

In [12]:
mean = df["Academic_Readiness_Score"].mean()
std = df["Academic_Readiness_Score"].std()

highest = df["Academic_Readiness_Score"].max()
lowest = df["Academic_Readiness_Score"].min()

z_high = (highest - mean) / std
z_low = (lowest - mean) / std

print("="*60)
print("NORMAL DISTRIBUTION")
print("="*60)

print("Mean:", round(mean,2))
print("Standard Deviation:", round(std,2))
print()

print("Highest Score:", highest)
print("Highest Z-score:", round(z_high,2))
print()

print("Lowest Score:", lowest)
print("Lowest Z-score:", round(z_low,2))

print("\nEmpirical Rule")
print("Within ±1 SD : 68%")
print("Within ±2 SD : 95%")
print("Within ±3 SD : 99.7%")

NORMAL DISTRIBUTION
Mean: 71.77
Standard Deviation: 9.69

Highest Score: 98.0
Highest Z-score: 2.71

Lowest Score: 41.9
Lowest Z-score: -3.08

Empirical Rule
Within ±1 SD : 68%
Within ±2 SD : 95%
Within ±3 SD : 99.7%


## Normal Distribution Analysis

The Academic Readiness Score has:

- Mean = **71.77**
- Standard Deviation = **9.69**

The highest score (**98.0**) has a Z-score of **2.71**, meaning it lies approximately **2.7 standard deviations above the mean**, making it an exceptionally high score.

The lowest score (**41.9**) has a Z-score of **−3.08**, indicating it is more than **3 standard deviations below the mean**, making it an unusually low score.

According to the Empirical Rule for a normal distribution:

- Approximately **68%** of observations are expected within **±1 standard deviation**.
- Approximately **95%** are expected within **±2 standard deviations**.
- Approximately **99.7%** are expected within **±3 standard deviations** of the mean.

# Final Statistical Observations

1. Students study an average of **15.69 hours per week**, while the average sleep duration is approximately **7 hours per day**, indicating balanced academic and personal routines.

2. Academic Readiness Score exhibits the greatest variability (SD = **9.69**), suggesting noticeable differences in students' preparedness levels.

3. Monthly Discretionary Spending contains the highest number of outliers among the selected variables, and removing these outliers reduced the mean more than the median, demonstrating the influence of extreme values.

4. Approximately **25.3%** of students have a part-time job, while **59.3%** exercise at least three days per week.

5. Students with part-time jobs appear more likely to experience high stress, as Events A and B were found to be **not independent**.

6. Bayes' Theorem produced the same conditional probability as the direct calculation, validating the probability analysis.

7. The highest Academic Readiness Score is about **2.71 standard deviations above the mean**, whereas the lowest score is about **3.08 standard deviations below the mean**, indicating a few exceptional observations at both extremes.

# Conclusion

This analysis applied descriptive statistics, measures of dispersion, outlier detection using the IQR method, probability concepts, conditional probability, Bayes' Theorem, and normal distribution analysis to the Student Wellbeing Survey dataset.

The results indicate moderate study hours, consistent sleep patterns, varying academic readiness, and the presence of several spending-related outliers. Probability analysis showed that part-time employment and high stress are associated rather than independent. Bayes' Theorem successfully verified the conditional probability calculation, and Z-score analysis identified exceptionally high and low academic readiness scores. Overall, the analysis provides meaningful statistical insights into student wellbeing and academic preparedness while demonstrating the practical application of fundamental statistical concepts.

In [13]:
df.to_csv("Student_Wellbeing_Cleaned.csv", index=False)